### PR TITLE
Add Proton GE

### PR DESCRIPTION
**Pull request by Lctrs**
Tested with Football Manager 2021, seems to work.

Should be merged in a `beta` branch.

Friendly ping @nanonyme and @gasinvein. You might be interested into reviewing this. :)

### CODE REVIEW

**Comment by gasinvein** [comment]
I'd prefer to keep common git history with original proton flatpak, if possible. It would be useful for porting changes back and forth.

**Comment by Lctrs** [comment]
@gasinvein Done.

**Comment by gasinvein** [comment]
@Lctrs Did you do a 3-way-merge or something? Github doesn't seem to recognize changes against 589f1fb

**Comment by Lctrs** [comment]
@gasinvein Fixed.

**Comment by gasinvein** [comment]
@Lctrs Thanks. You can also use symlinks to avoid renaming the manifest and appdata, yet I'm not sure if flathub infra will handle it fine.

File: patches/proton/keep-system-gstreamer-plugins.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/proton b/proton
+index ea6e9af..17ce908 100755
+--- a/proton
++++ b/proton
+@@ -607,7 +607,7 @@ class Session:
+ 
+         self.env["WINEDLLPATH"] = g_proton.lib64_dir + "/wine:" + g_proton.lib_dir + "/wine"
+ 
+-        self.env["GST_PLUGIN_SYSTEM_PATH_1_0"] = g_proton.lib64_dir + "gstreamer-1.0" + ":" + g_proton.lib_dir + "gstreamer-1.0"
++        self.env["GST_PLUGIN_PATH"] = g_proton.lib64_dir + "gstreamer-1.0" + ":" + g_proton.lib_dir + "gstreamer-1.0"
```
Comment (gasinvein) [inline patches/proton/keep-system-gstreamer-plugins.patch:10]: This patch isn't needed since you're bundling gstreamer.

**Comment by Lctrs** [comment]
@gasinvein Let's find out.

File: patches/proton/keep-system-gstreamer-plugins.patch (hunk 1 above)
Comment (Lctrs) [reply patches/proton/keep-system-gstreamer-plugins.patch]: Adressed.

**Comment by gasinvein** [comment]
I guess bundling gstreamer may be more tricky than it seems. We'll need to make sure that gst-plugins pick up proper (bundled) gstreamer via pkgconfig, and so do FAudio and Wine. Also RPATH probably need to be set so that proper gstreamer libs are loaded at run time.

**Comment by Lctrs** [comment]
How can I check this ?

**Comment by gasinvein** [comment]
Good question. At run-time, you can use `ldd` on faudio/wine libs. At build time, I guess the easiest way is to run a shell in flatpak-builder sandbox with `--build-shell=gst-plugins-base` and manually check which pkgconfig file was selected with `pkg-config --path gstreamer-1.0`

**Comment by Lctrs** [comment]
OK they effectively pick system gstreamer (and not the bundled one).

**Comment by gasinvein** [comment]
@Lctrs Submitted flathub/com.valvesoftware.Steam.CompatibilityTool.Proton#22. It may help with bundling gstreamer.

**Comment by Lctrs** [comment]
@gasinvein You're a beast ! Everything is good now.

**Comment by Lctrs** [comment]
bot, build com.valvesoftware.Steam.CompatibilityTool.ProtonGECustom

**Comment by Lctrs** [comment]
@gasinvein looks like flathub infra supports symlinks :)

**Comment by gasinvein** [comment]
Ok, so, to keep git difference at minimum, before it's merged, I'd suggest to do the following:
- Rebase on flathub/com.valvesoftware.Steam.CompatibilityTool.Proton@c430b5b and drop commits 6c08a12 and 387212b
- Squash commits 0b9d1e6 and 33aa754

**Comment by Lctrs** [comment]
@gasinvein Done.

**Comment by Lctrs** [comment]
bot, build com.valvesoftware.Steam.CompatibilityTool.ProtonGECustom

File: com.valvesoftware.Steam.CompatibilityTool.Proton.yml [hunk 2]
```diff
@@ -13,43 +13,62 @@ sdk-extensions:
...
+  type: file
+  url: https://raw.githubusercontent.com/KhronosGroup/Vulkan-Docs/v1.2.160/xml/vk.xml
+  sha256: 73b449328899fb67703e9a229b6c8858f31b2933485b08710b4f69f842add94b
+
+x-proton-patches: &proton_patches
+  type: shell
+  commands:
+    - |
+      sed \
+        -e "s#\./dlls/winevulkan/make_vulkan#./dlls/winevulkan/make_vulkan --verbose --xml $(pwd)/vk.xml#g" \
```
Comment (gasinvein) [inline com.valvesoftware.Steam.CompatibilityTool.Proton.yml:64]: Maybe just put `vk.xml` to `wine/dlls/winevulkan/vk-1.2.160.xml` (or wherever `make_vulkan` expects it to be) instead editing the script?

Comment (Lctrs) [reply com.valvesoftware.Steam.CompatibilityTool.Proton.yml:64]: I would like to have the verbose flag as well.

File: compatibilitytool.vdf [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"compatibilitytools"
+{
+  "compat_tools"
+  {
+    "proton-ge-custom_flatpak"
+    {
+      "display_name" "Proton GE Custom @VERSION@ (community build @BUILD@)"
```
Comment (gasinvein) [inline compatibilitytool.vdf:7]: I think "community" doesn't make much sense here, since Proton-GE itself isn't an "official" thing. Ping @nanonyme

**Comment by gasinvein** [comment]
This takes ridiculous amount of time to build, more than Chromium, it's even worse than normal Proton (which also builds much slower than it should). I believe this is due to enormous amount of git fetches/checkouts. We should do something about it.
My initial idea was to pack upstream sources to a tarball externally, then use that tarball as archive type source, yet I didn't go into details, namely where the tarball should be created and stored. Any suggestions? Maybe ask upstream to provide source tarball (with patches applied)?

**Comment by Lctrs** [comment]
Will contact upstream about this.

**Comment by Lctrs** [comment]
I don't know if that's feasible or not but it could also be an idea of improvement for `flatpak-builder` to mutualise sources operations across modules when they share same sources definitions.

**Comment by gasinvein** [comment]
As a workaround, we can also use single git repo per flatpak-builder module instead of checking out the whole Proton-GE repo each time, but this would require much more handwork.

**Comment by Lctrs** [comment]
No response from upstream yet. Anyway, I created https://github.com/Lctrs/proton-ge-custom-tarball-maker, a simple Github Actions to create source tarball from Proton GE patched sources. Let's see how it reduces build time.

bot, build com.valvesoftware.Steam.CompatibilityTool.ProtonGECustom

File: compatibilitytool.vdf (hunk 3 above)
Comment (Lctrs) [reply compatibilitytool.vdf:7]: Well, it's a community build of a community Proton distribution :smile:

**Comment by gasinvein** [comment]
@Lctrs vkd3d won't build outside of a git repo. I've previously made a fix for this gasinvein/proton-flatpak@9b453c4b808c11ea090e1e87444dc7ee508a6f9f

**Comment by Lctrs** [comment]
bot, build com.valvesoftware.Steam.CompatibilityTool.ProtonGECustom

**Comment by barthalion** [comment]
No complaints from me, but as it's not my area, I'll wait for approvals from @gasinvein and @nanonyme.

**Comment by gasinvein** [comment]
@barthalion Do you have an opinion on using custom source tarball? IMO it adds complexity and isn't not good for transparency, but it reduces build time from ~3 hours to ~40 minutes.

File: compatibilitytool.vdf (hunk 3 above)
Comment (gasinvein) [reply compatibilitytool.vdf:7]: Also, won't the name be too long? Is the version still visible in Steam UI?

Comment (Lctrs) [reply compatibilitytool.vdf:7]: Yep.

Capture d’écran de 2020-12-01 15-02-53

**Comment by Lctrs** [comment]
@gasinvein I have inlined flatpak's id and display name with what you did in https://github.com/gasinvein/proton-flatpak.

**Comment by gasinvein** [comment]
@Lctrs That wasn't necessary, I've did some PoC there that shouldn't be considered entirely correct. But well, I guess it'll be easier to sync changes now.

**Comment by barthalion** [comment]
> @barthalion Do you have an opinion on using custom source tarball? IMO it isn't good for transparency, but it reduces build time from ~3 hours to ~40 minutes.

Yeah, it's not ideal but not completely uncommon in the packaging world either so let's keep it.

**Comment by Lctrs** [comment]
Looks like the merge action doesn't like symlinks : https://github.com/flathub/flathub/runs/1491179128

**Pull request closed without merging**